examples/fabric/byot/trainer.py

```diff
@@ -0,0 +1,409 @@
+
+        remainder = self.fabric.load(path, state)
+        self.global_step = remainder.pop("global_step")
+        self.current_epoch = remainder.pop("current_epoch")
+
+        if remainder:
+            raise RuntimeError(f"Unused Checkpoint Values: {remainder}")
+
+    def save(self, state):
+        if state is None:
+            state = {}
+
+        state.update(global_step=self.global_step, current_epoch=self.current_epoch)
+
+        self.fabric.save(os.path.join(self.checkpoint_dir, f"epoch-{self.current_epoch:04d}.ckpt"), state)
+
+    @staticmethod
+    def get_latest_checkpoint(checkpoint_dir):
+        if not os.path.isdir(checkpoint_dir):
+            return None
+
+        items = sorted(os.listdir(checkpoint_dir))
+        return os.path.join(checkpoint_dir, items[-1])
+
+    def _parse_optimizers_schedulers(
+        self, configure_optim_output
+    ) -> Tuple[Optional[Optimizable], Optional[Mapping[str, Union[LRScheduler, bool, str, int]]]]:
+
+        _lr_sched_defaults = {"interval": "epoch", "frequency": 1, "monitor": "val_loss"}
+
+        # single optimizer
+        if isinstance(configure_optim_output, Optimizable):
+            return configure_optim_output, None
+
+        # single lr scheduler
+        elif isinstance(configure_optim_output, LRScheduler):
+            return None, _lr_sched_defaults.update(scheduler=configure_optim_output)
+
+        # single lr scheduler config
+        elif isinstance(configure_optim_output, Mapping):
+            _lr_sched_defaults.update(configure_optim_output)
+            return None, _lr_sched_defaults
+
+        # list or tuple
+        elif isinstance(configure_optim_output, (list, tuple)):
+            if all([isinstance(_opt_cand, Optimizable) for _opt_cand in configure_optim_output]):
+                # single optimizer in list
+                if len(configure_optim_output) == 1:
+                    return configure_optim_output[0][0], None
+
+                raise NotImplementedError("BYOT only supports a single optimizer")
+
+            elif all([isinstance(_lr_cand, (LRScheduler, Mapping)) for _lr_cand in configure_optim_output]):
+
+                # single scheduler in list
+                if len(configure_optim_output) == 1:
+                    return None, self._parse_optimizers_schedulers(configure_optim_output[0])[1]
+
+            # optimizer and lr scheduler
+            elif len(configure_optim_output) == 2:
+                opt_cands, lr_cands = (
+                    self._parse_optimizers_schedulers(configure_optim_output[0])[0],
+                    self._parse_optimizers_schedulers(configure_optim_output[1])[1],
+                )
+                return opt_cands, lr_cands
+
+        return None, None
+
+    @staticmethod
+    def _format_iterable(
+        prog_bar, candidates: Optional[Union[torch.Tensor, Mapping[str, Union[torch.Tensor, float, int]]]], prefix: str
+    ):
+        if isinstance(prog_bar, tqdm) and candidates is not None:
```

Comment:
Should we use rich instead?

Reply:
I thought about this found tqdm to be easier to configure and better known in general. Open for changes though